File: algorithms/code/leetcode/lc203_remove_linked_list_elements/LC203RemoveLinkedListElements.java
package leetcode.lc203_remove_linked_list_elements;

/**
 * https://leetcode.com/problems/remove-linked-list-elements/ #easy
 */
public final class LC203RemoveLinkedListElements {
    public static class ListNode {
        public int val;
        public ListNode next;

        ListNode(final int x) {
            val = x;
        }
    }

    public ListNode removeElements(final ListNode headNode, final int val) {
        ListNode head = headNode;
        if (head == null) {
            return null;
        }
        ListNode previous = null;
        ListNode node = head;
        while (node != null) {
            if (node.val == val) {
                if (previous == null) {
                    head = node.next;
                } else {
                    previous.next = node.next;
                }
            } else {
                previous = node;
            }
            node = node.next;
        }
        return head;
    }
}
package leetcode.lc203_remove_linked_list_elements;

import org.junit.jupiter.api.Test;

import leetcode.lc203_remove_linked_list_elements.LC203RemoveLinkedListElements.ListNode;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

public final class LC203RemoveLinkedListElementsTests {
    public ListNode convert(final int[] array) {
        ListNode list = null;
        ListNode last = null;
        for (int value : array) {
            ListNode node = new ListNode(value);
            node.next = null;
            if (list == null) {
                list = node;
            } else {
                last.next = node;
            }
            last = node;
        }
        return list;
    }

    @Test
    public void test612345676And6() throws Exception {
        int[] array = { 6, 1, 2, 3, 4, 5, 6, 7, 6 };
        ListNode list = convert(array);
        list = new LC203RemoveLinkedListElements().removeElements(list, 6);
        int[] expected = { 1, 2, 3, 4, 5, 7 };
        for (int value : expected) {
            assertEquals(value, list.val);
            list = list.next;
        }
    }

    @Test
    public void testNothing() throws Exception {
        assertNull(new LC203RemoveLinkedListElements().removeElements(null, 0));
    }
}

File: algorithms/code/leetcode/lc203_remove_linked_list_elements/lc203_remove_linked_list_elements.py
# https://leetcode.com/problems/remove-linked-list-elements/

from __future__ import annotations

import unittest


class ListNode:
    def __init__(self, x: int) -> None:
        self.val: int = x
        self.next: ListNode | None = None


class Solution:
    def removeElements(
        self, head: ListNode | None, val: int
    ) -> ListNode | None:
        if head is None:
            return None
        previous: ListNode | None = None
        node: ListNode | None = head
        while node is not None:
            if node.val == val:
                if previous is None:
                    head = node.next
                else:
                    previous.next = node.next
            else:
                previous = node
            node = node.next
        return head


class TestCode(unittest.TestCase):
    def convert(self, array: list[int]) -> ListNode | None:
        handle = ListNode(0)
        last = handle
        for value in array:
            node = ListNode(value)
            last.next = node
            last = node
        return handle.next

    def test_6_1_2_3_4_5_6_7_6__6(self) -> None:
        array = [6, 1, 2, 3, 4, 5, 6, 7, 6]
        listed = self.convert(array)
        listed = Solution().removeElements(listed, 6)
        expected = [1, 2, 3, 4, 5, 7]
        for value in expected:
            assert listed is not None
            assert value == listed.val
            listed = listed.next

    def test_nothing(self) -> None:
        assert Solution().removeElements(None, 0) is None
